maven/lib/dependabot/maven/utils/auth_details_finder.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Dependabot
+  module Maven
+    module Utils
+      class AuthDetailsFinder
+
+        def initialize(credentials)
+          @credentials = credentials
+        end
+
+        def auth_details(maven_repo_url)
+          cred =
+            credentials.select { |c| c["type"] == "maven_repository" }.
+            find do |c|
+              cred_url = c.fetch("url").gsub(%r{/+$}, "")
+              next false unless cred_url == maven_repo_url
+
+              c.fetch("username", nil)
+            end
+
+          return gitlab_auth_details(maven_repo_url) unless cred
+
+          token = cred.fetch("username") + ":" + cred.fetch("password")
+          encoded_token = Base64.encode64(token).delete("\n")
+          { "Authorization" => "Basic #{encoded_token}" }
+        end
+
+        private
+
+        attr_reader :credentials
+
+        def gitlab_auth_details(maven_repo_url)
+          return {} unless gitlab_maven_repo?(URI(maven_repo_url).path)
+
+          cred =
+            credentials.select { |c| c["type"] == "git_source" }.
+            find do |c|
+              cred_host = c.fetch("host").gsub(%r{/+$}, "")
+              next false unless URI(maven_repo_url).host == cred_host
+
+              c.fetch("password", nil)
+            end
+
+          return {} unless cred
+
+          { "Private-Token" => cred.fetch("password") }
+        end
+
+        def gitlab_maven_repo?(maven_repo_path)
+          gitlab_maven_repo_reg = %r{^(/api/v4).*(/packages/maven)/?$}.freeze
```

Comment:
It doesn't look like we use the capture groups. Do this need them?

Should `.*` become `.+/-`. I think this is meant to capture the group hierarchy in gitlab projects. E.g `org/group/subgroup/project`. Is that right?

Reply:
You are right, I will remove them.

The intention was to focus on the start and the end of the path, and allow anything or nothing in between, as the path for the same repo can be on group level, project level or instance level: https://docs.gitlab.com/ee/user/packages/maven_repository/#use-the-gitlab-endpoint-for-maven-packages